Implement scroll, scrollLeft, scrollTop and friends, addressing issue #9650

izgzhen opened the pull request:
This is a work to solve https://github.com/servo/servo/issues/9650. Thanks a lot for helping the review.
- [x] scroll
- [x] scrollTo
- [x] scrollBy
- [x] scrollTop (setter and getter)
- [x] scrollLeft (setter and getter)

The setters will be implemented in another PR after this is merged.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/9968)

<!-- Reviewable:end -->

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @kichjang (or someone else) soon.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

asajeffrey commented:
Hi there @izgzhen, thanks for contributing!

@danlrobertson, does this PR overlap with #9824 at all?

danlrobertson commented:
The spec is similar, but there shouldn't be too much overlap. I wanted to implement `scrollWidth` and `scrollHeight` first so we could create a method for the [`scrolling area`](https://drafts.csswg.org/cssom-view/#scrolling-area), which is needed for a few of the methods here, but that is the only overlap I can see at the moment.

@izgzhen: I think you're missing `SetScrollLeft` in your checklist. Great work!

izgzhen commented:
@asajeffrey There is something different, and I will try to avoid duplication by abstracting common utilities out.

@danlrobertson thanks, I don't write it on the list because I haven't write a single line of code yet :P

asajeffrey commented:
#9824 landed!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #9967) made this pull request unmergeable. Please resolve the merge conflicts.

asajeffrey commented:
@izgzhen would you like me to start reviewing, or hold off for the moment?

izgzhen commented:
@asajeffrey Please wait for a minute, I will need a little time to rebase and think about incorporating changes introduced in PR #9824.

izgzhen commented:
@asajeffrey rebase and some more polishing is done, I am hitting the wall now :P

I have two problems now:

1. How to get the computed style for a given element (i.e. what `get_computed` does)? The motivation is to "check whether the element has [any associated CSS layout box](https://drafts.csswg.org/cssom-view/#css-layout-box), and whether the element has [any associated scrolling box](https://drafts.csswg.org/cssom-view/#scrolling-box). The pull #9968 did a similar thing through `layout_rpc` and process the query in `layout::query`, but I don't know it is also proper to my case and if there is an *easier* way to complete this.
2. I intend to use the `get_scroll_area` by pull #9968 at the final line of getting on ScrollLeft and ScrollTop, but first there is a coercion, and second some checking over document and window *did* overlap, although [the spec of ScrollLeft and ScrollWidth](https://drafts.csswg.org/cssom-view/#dom-element-scrollleft) differ in a very interleaving way :(

asajeffrey commented:
I did some preliminary reading through, which makes me think that there needs to be a bit of care about what is handled by the script thread, and what is handled by the layout thread (c.f. the approach in #9824). I think you're going to need to construct a message which is sent from script to layout requesting the necessary calculated values, and there should be at most one such message for each call to a DOM method.

@danlrobertson: you recently looked at this code, care to weigh in?

---

Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/element.rs, line 173 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9968#-KCcXM6k1q711-zpztmK:-KCcXM6lh7lc9_UGtBgy:-914000975) ([raw file](https://github.com/servo/servo/blob/a066675cffedc809ce35fd83ffa500c8e6a32b90/components/script/dom/element.rs#L173)):*
I think "elements that have a computed value of the display property that is table-column or table-column-group must be considered to have an associated CSS layout box" is intended as a comment, not a definition. Indeed, the spec says "ISSUE 1: The terms CSS layout box and SVG layout box are not currently defined by CSS or SVG."

For the moment I think this function should just return `true`, but we should probably check on IRC that this is correct. If so, we should file an issue saying that the spec is unclear here.

---

*[components/script/dom/element.rs, line 180 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9968#-KCcYBkl3KBlOSe2tmIV:-KCcYBkl3KBlOSe2tmIW:-307685226) ([raw file](https://github.com/servo/servo/blob/a066675cffedc809ce35fd83ffa500c8e6a32b90/components/script/dom/element.rs#L180)):*
I think this can be simplified to:

* The element has an associated CSS layout box, and
* The element’s used value of the overflow-x or overflow-y properties is not visible.

This is the first method that requires computed values.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9968#-:-KCca6eF5Ozd8muh1VYd:540360942)*
<!-- Sent from Reviewable.io -->

danlrobertson commented:
Agreed. A good portion of this will definitely involve the layout thread. I'll do some more reading as well and come back with hopefully more info.

> did overlap, although the spec of ScrollLeft and ScrollWidth differ in a very interleaving way :(

Yeah, I noticed there were a couple very small differences (for the getters). I wonder how much of it is just wording and how much of it is actually different. I can ask on IRC and hope for some responses. Which parts in particular are you talking about?

Just a suggestion. For these particular methods I would focus on implementing the "working" or "hard" part of the algorithm first. In your case the last step of each method. Furthermore, I would work on the setters first. It's a bit complicated as you'll have to perform a scroll (not sure if it has been implemented yet), but once you have that, i think you might have a better foundation to work off of and it should be easier to go back and implement the restrictions. Just a suggestion... feel free not to take it :smile:

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/element.rs, line 180 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9968#-KCcYBkl3KBlOSe2tmIV:-KCcwBvhA5Cz-v5OAZ__:-334669205) ([raw file](https://github.com/servo/servo/blob/a066675cffedc809ce35fd83ffa500c8e6a32b90/components/script/dom/element.rs#L180)):*
You could use `Fragment::style` to get `ComputedValues::get_box` to get the [`Box`](https://doc.servo.org/style/properties/style_structs/struct.Box.html) which contains `overflow_x` and `overflow_y`, but I'm not sure if that can be accessed outside of the layout thread (correct me if I'm wrong). 

---

*[components/script/dom/element.rs, line 1534 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9968#-KCcwv3iQRUABcixCtvA:-KCcwv3jnl21JRh08SXv:1258468092) ([raw file](https://github.com/servo/servo/blob/a066675cffedc809ce35fd83ffa500c8e6a32b90/components/script/dom/element.rs#L1534)):*
Is this preferred or is `quirks_mode() == Quirks`? This will allow `LimmitedQuirks`

---

*[components/script/dom/element.rs, line 1545 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9968#-KCcxBOpvX3l7hKvYUiN:-KCcxBOpvX3l7hKvYUiO:1662569705) ([raw file](https://github.com/servo/servo/blob/a066675cffedc809ce35fd83ffa500c8e6a32b90/components/script/dom/element.rs#L1545)):*
Same here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9968#-:-KCcvbAp-Yhx9lkYa8-J:-822679309)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/element.rs, line 1534 \[r4\]](https://reviewable.io:443/reviews/servo/servo/9968#-KCcwv3iQRUABcixCtvA:-KCd697shzS90aUJZxyB:-1470161246) ([raw file](https://github.com/servo/servo/blob/a066675cffedc809ce35fd83ffa500c8e6a32b90/components/script/dom/element.rs#L1534)):*
It sounds sensible. I should take spec more literally.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
@izgzhen would you like me to dive into the current code base some more, or hold off? It sounds like we're in agreement that the code is going to need quite a bit of re-architecting to move computation from the script thread to the layout thread.

izgzhen commented:
@asajeffrey You idea sounds great. Hope I can be helpful to you.

asajeffrey commented:
@izgzhen is this PR still active?

izgzhen commented:
@asajeffrey Yes, I believe so. Sorry that I have been busy with GSoC proposal and coursework these days. Also, I am kind of waiting more comments so I procrastinated :(

If there is not major problem about "moving computation from the script thread to the layout thread", I am going to review the code and write some test cases this weekend.

izgzhen commented:
@asajeffrey hi, I have investigated and experimented throughly this weekend, but the result turns out to be not very pleasing.

First, the `origin.x` and `origin.y` is implemented as `window.ScrollX()` and `window.ScrollY()` in PR #9824, which seems not to be what we want. Also, I take a look at the code about scrolling in compositor and layout, and it looks horribly complex here. `pub scroll_offset: TypedPoint2D<LayerPixel, f32>` field in `components/compositing/compositor_layer.rs`, line 46 seems related. I think I should somehow get the value from here and send it to the script thread. Also, I might need to call the interfaces there to perform setting on `scrollTop` and friends.

Second, a lot of other scrolling utilities are not supported in `Element` interface, such as `scroll`. It might not be a big problem now, but I think it is hard to implement one without referring to another.

As a conclusion, I think I am going to investigate a bit deeper and try to give a more complete solution. It is very rewarding process to understand and identify the problem, although there is not any fruit yet. I thin k we can leave this PR open here as a place for discussion and experimentation, but I am afraid it might linger a very long time. What is your idea?

danlrobertson commented:
> First, the origin.x and origin.y is implemented as window.ScrollX() and window.ScrollY() in PR #9824, which seems not to be what we want

`get_scroll_area` will return the [scroll area](https://drafts.csswg.org/cssom-view/#scrolling-area). The `origin` being

> the top left padding edge of the element when the element has its default scroll position...

Now that I'm looking at the spec again, there is a chance that our current implementation might need some attention once `scroll` is implemented due to "when the element has its default scroll position".

> Second, a lot of other scrolling utilities are not supported in Element interface, such as scroll.

`scroll` is definitely not implemented yet, and will need to be implemented to correctly implement `SetScroll(Top|Left)`. Maybe `handle_scroll_event` would be a good place to start? I don't know much about scrolling, but wouldn't mind looking into it more if you could use some help.

izgzhen commented:
Here are some notes of my latest "discoveries":

First, as stated a week ago, the `scroll_offset` of `extra_data` field of `CompositorData` is what we want. As tested, the `handle_scroll_event` will trigger the updating of this field. Note that, the data is type parameter to an abstract `Layer` as defined in `layers` crate. However, the actual layer is identified by `LayerId` as declared in `gfx_traits`. As a result, we need two id, pipeline id and layer id, to get the right data from compositor layer.

Second, the `compositor_thread` defines some messages. The `ScriptToCompositor` message is one we are interested in. We can send the extended message from window, since it has a compositor IPC threads field.

The final problem is how to relate element to such a layer. As tested with `tests/html/simple-overflow-scroll.html`, the document viewport's pipeline and the section element's pipeline is same, while they have different layer id: Document one is `0-FragmentBody-0`, and section element one is `4734743072-FragmentBody-0`. So we can assume that there is kind of relationship. Actually, the `Fragment` as defined in `layout` component has two methods: `fn layer_id(&self) -> LayerId` and `fn layer_id_for_overflow_scroll(&self) -> LayerId`. This is exactly what we want!

The above is idea for getting, and setting should be similar (calling `handle_scroll_event` seems feasible).

Up the implementation, I think I can reuse the `UnioningFragmentScrollAreaIterator` as contributed by @danlrobertson by adding a `layer_id` field. And back to window, the `scroll_area_query` can be inspect the `layer_id` from layout RPC, send with its own `pipeline_id` to the compositor thread (XXX: Hoping it is the right pipeline, I am kinda arbitrary here). Then, we can get the `scroll_offset`, which is the `ScrollTop` and `ScrollLeft` we want. This is a quick hack I think.

Some other consideration is to make a new fragment iterator getting solely `LayerId`, so it might be reused in further as well. And on the query side, we will wrap it add getter and setter for `scroll_offset`, and deal with further semantics inside the script thread.

Plus the existing hack to retrieve the computed value of `overflow-x` and `overflow-y`, this sounds a bit of engineering. Welcome comments!

izgzhen committed:
Implement ScrollTop and ScrollLeft getters:

Add new compositor message to get scroll_offset;
Add new layout query for computed value of overflow-x/y;
Implement layer_id method for ThreadSafeLayoutNode;
Add new layout query for layer_id;
Implement script interface for getting scrollTop and scrollLeft, as well as relavant helper functions.

izgzhen commented:
In the new commits, I finally fetched the right `layer_id` and sent the `scroll_offset` back to the script thread. However, it still has two problems:

1. The correct `layer_id` is the first one in the process of integrating the fragments (https://github.com/servo/servo/pull/9968/commits/74695da5bec09438158fe6cfd817068af5b6713b#diff-8002b77bfa2ea6411d84e780a9955b85R438). But this hack is based on observable test results, rather than reasoning.
2. The returned value (`scroll_offset`) has one border-width difference from WebKit and Chrome testing. But since this is an offset (i.e. a relative distance), I can't simply add the border-width on it. I suspect that the scroll_offset implemented in Servo previously has subtle semantics difference from spec.

Despite these issues, the current code is basically working as intended. Welcome reviewing.

jdm commented:
Well done! :dancer:

izgzhen commented:
@asajeffrey Got time to review the changes? I have clean up the warnings and this is the best version I can give now. After discussing the two issues as stated above and getting this patch merged, I am planning to open another PR to implement the element `scroll` and its setters friends.

asajeffrey commented:
Thanks! I'll have a look at the commits and get back to you.

asajeffrey commented:
@danlrobertson looking at #9824 again, line 425 of `components/layout/query` is (https://github.com/servo/servo/pull/9824/files#diff-8002b77bfa2ea6411d84e780a9955b85R425):
```rust
    self.origin_rect = Rect::new(Point2D::new(top_padding, left_padding),
                                 Size2D::new(right_padding, bottom_padding));
```
Shouldn't that be:
```rust
    self.origin_rect = Rect::new(Point2D::new(left_padding, top_padding),
                                 Size2D::new(right_padding, bottom_padding));
```
that is, the x and y coordinates of the origin are swapped? This might explain some of the discrepancies we're seeing.

danlrobertson commented:
Yes, that should most definitely be swapped.

asajeffrey commented:
Thanks! The overall architecture looks good, the only big picture question I have about this patch is about getting the layer_id, see the detailed comments below.

Sorry about the delay in replying, I've been in a 2-day meeting!

---

Reviewed 9 of 11 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[components/layout/query.rs, line 122 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnE2iUlNBJC6TA7w9n:-KEnE2iUlNBJC6TA7w9o:857763337) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L122)):*
Can't this just be
```rust
    match self.0.lock().unwrap().overflow_response {
            NodeOverflowResponse(x) => NodeOverflowResponse(x),
    }
```

---

*[components/layout/query.rs, line 131 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnERLY8XXrrLTldGoP:-KEnERLY8XXrrLTldGoQ:-1298117883) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L131)):*
Can the second `unwrap()` fail?

---

*[components/layout/query.rs, line 362 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnEdGZDhd_98QU4L2L:-KEnEdGZDhd_98QU4L2M:1664865038) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L362)):*
I'm not sure that repurposing `UnioningFragmentScrollAreaItertator` is the right way to implement a getter for the layer id.

---

*[components/layout/query.rs, line 449 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnEu76oJdOCOwwC7U9:-KEnEu76oJdOCOwwC7UA:-775780565) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L449)):*
If it's the first layer id, why do we need an iterator?

---

*[components/layout/query.rs, line 545 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnF5-0cZhJDc-5bR-u:-KEnF5-0cZhJDc-5bR-v:2060971463) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L545)):*
This is potentially having to iterate through a lot of fragments, I'm concerned about the performance.

---

*[components/script/layout_interface.rs, line 144 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnFMIL8WL8cd2xrF0z:-KEnFMIL8WL8cd2xrF1-:-482076794) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/script/layout_interface.rs#L144)):*
Is there an alternative way to specify the contained type. Should this be a `Point2D`?

---

*[components/script/dom/element.rs, line 191 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnGTzXFDqkT9K1cRMh:-KEnGTzXFDqkT9K1cRMi:-1555507689) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/script/dom/element.rs#L191)):*
Can we implement `overflow_visible` rather than its negation? Often positive code is easier to read than negative code.

---

*[components/script/dom/element.rs, line 194 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnHLHSUA-4r2Ch5p9r:-KEnHLHSUA-4r2Ch5p9s:1369144769) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/script/dom/element.rs#L194)):*
Can the methods `overflow_x_is_visible` and `overflow_y_is_visible` be used?

---

*[components/script/dom/window.rs, line 1128 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnHobUEwMdMsC5HhCT:-KEnHobUEwMdMsC5HhCU:1114735890) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/script/dom/window.rs#L1128)):*
Can we return a Box or a Point2D?

---

*[components/script/dom/window.rs, line 1144 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnHziO7SZMEwRtKvAv:-KEnHziPdW7roY1K2IT0:213337246) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/script/dom/window.rs#L1144)):*
I wonder if there's a way to do this without having to get the layer_id from layout before getting the offset from the compositor?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968#-:-KEnIRve8DK5qka4VjCG:1164943627)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[components/layout/query.rs, line 122 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnE2iUlNBJC6TA7w9n:-KEnK7gWmoHut46Kkrcj:-360064374) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L122)):*
Yes, thanks for reminding :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[components/layout/query.rs, line 131 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnERLY8XXrrLTldGoP:-KEnKUkyOn8DHyibcAHz:1217694479) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L131)):*
I think it can't fail since before it is returned, we process the request and iterate the fragments. There should be at least one fragment (Am I correct here?), so

        if self.layer_id == None {
            self.layer_id = Some(fragment.layer_id());
        }

in `impl FragmentBorderBoxIterator for UnioningFragmentScrollAreaIterator` will write the `layer_id` into a Some value.

The `Option` is used because we need to initialize the field into something sensible.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[components/layout/query.rs, line 362 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnEdGZDhd_98QU4L2L:-KEnMTaPbA2XH_S1t68W:-1409860970) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L362)):*
I have to admit that I'm kinda lazy and tend to reuse the code before I'm sure if it is necessary to add another iterator for this particular purpose :P

What's your idea about writing another iterator?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[components/script/dom/window.rs, line 1144 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnHziO7SZMEwRtKvAv:-KEnQIWCjpO_EDNxVbNa:-1723824211) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/script/dom/window.rs#L1144)):*
I wonder this as well. The `scroll_offset` in layer are indexed by `layer_id` and `pipeline_id` with `IOCompositor::find_layer_with_pipeline_and_layer_id`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[components/script/dom/element.rs, line 194 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnHLHSUA-4r2Ch5p9r:-KEnRTOPCyU59K7IIPEO:454898101) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/script/dom/element.rs#L194)):*
I batched them together due to performance concern. Does my concern make sense here? Finer-grained query seems more reusable though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

danlrobertson commented:
Sorry, I got distracted and forgot to actually post my comments :smile:

I'm doing some digging and LayoutId is constructed with an `OpaqueNode`, which we have from the start, and the `PseudoElementType`. Checkout the code for `Fragment::layout_id`. We would still need `Fragment::pseudo`, so we still need to find the `Fragment`.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[components/layout/query.rs, line 362 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnEdGZDhd_98QU4L2L:-KEnPkEBaRto3kUnaseu:162745307) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L362)):*
I definitely would not include `layer_id` in `UnioningFragmentScrollAreaIterator`. You will end up `process`ing all of the children. for simple tests like.

```html
   <div id="parent">
       <div id = "child">
       ...
       </div>
   </div>
```

Its not a big deal, but if you have a larger number of children, you'll end up processing a lot more than you need to.

---

*[components/layout/query.rs, line 448 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnQkovU54VewRYIs5M:-KEnQkovU54VewRYIs5N:1176366850) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L448)):*
So the `layer_id` you want is always going to be the first `layer_id` you come across. See the the `should_process` condition. Specifically

```rust
// will return true first when fragment.node == self.node_address
fragment.contains_node(self.node_address)...
// the rest of the iterations rely on the second condition. You don't want those
```

Note: After looking at this perhaps it would be easier to read if this was changed to `fragment.node == self.node_address...`

So if you're going the iterator route, definitely create a new one and 

---

*[components/script/dom/element.rs, line 1613 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9968#-KCuKvfQFjyNUCZIolAe:-KCuKvfRz6oOfdaLEDDX:58272226) ([raw file](https://github.com/servo/servo/blob/26d75f757b87bda34a74effe1d58be80cff314fd/components/script/dom/element.rs#L1613)):*
Shouldn't this follow the [scroll the element](https://drafts.csswg.org/cssom-view/#scroll-an-element) spec?

---

*[components/script/dom/window.rs, line 1144 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnHziO7SZMEwRtKvAv:-KEnQZpqGvrTMq34i_pl:-986085584) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/script/dom/window.rs#L1144)):*
I agree iterating seems kind of silly. I'll poke around a bit and see what I can find.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968#-:-KEnVm3i1hFSad7waQyN:-351704010)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[components/script/dom/element.rs, line 1613 \[r5\]](https://reviewable.io:443/reviews/servo/servo/9968#-KCuKvfQFjyNUCZIolAe:-KEnf64JlkXdDeC0Mddx:-1286546448) ([raw file](https://github.com/servo/servo/blob/26d75f757b87bda34a74effe1d58be80cff314fd/components/script/dom/element.rs#L1613)):*
This is the old experimental code actually. In the latest revision this should be dropped and I plan to implement setters in another PR. But thanks for concern though :D

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

danlrobertson commented:
Woohoo! No iteration! I'll take a longer look later.

---

Review status: 6 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[components/layout/query.rs, line 536 \[r8\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEpYM_bgxnhQefWDgjd:-KEpYM_bgxnhQefWDgje:402150012) ([raw file](https://github.com/servo/servo/blob/c4ccde386bf555757194b209b1de08a7b787765f/components/layout/query.rs#L536)):*
Any particular reason why this is an `Option`? We just `unwrap` it in `node_layer_id`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968#-:-KEpXqq6XXqp9gcqq7Zu:-411832364)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: 6 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[components/layout/query.rs, line 536 \[r8\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEpYM_bgxnhQefWDgjd:-KEpZzC2p7-DUNaVxTba:1395712676) ([raw file](https://github.com/servo/servo/blob/c4ccde386bf555757194b209b1de08a7b787765f/components/layout/query.rs#L536)):*
In line 1200 of layout_thread.rs:

```rust
ReflowQueryType::NodeLayerIdQuery(node) => {
    let node = unsafe { ServoLayoutNode::new(&node) };
    rw_data.layer_id_response = process_node_layer_id_request(node);
},
```

The response is a field of `LayoutThreadData`, which forces us to consider the initialization of this field. And I can only initialize it to `None` as a sensible choice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Looks a lot better now! Some comments on the code...

---

Reviewed 4 of 6 files at r8.
Review status: 10 of 12 files reviewed at latest revision, 6 unresolved discussions.

---

*[components/layout/query.rs, line 131 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnERLY8XXrrLTldGoP:-KEqMTQuC31hxGFpQTSE:1693655867) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L131)):*
Might be worth replacing unwrap by expect just in case there is no Fragment.

---

*[components/layout/query.rs, line 362 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnEdGZDhd_98QU4L2L:-KEqKHAjB213V6d7ASH2:2003583000) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/layout/query.rs#L362)):*
Yay, this is gone now!

---

*[components/layout/query.rs, line 441 \[r8\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEqKZ9LnZLS1bGIerI7:-KEqKZ9LnZLS1bGIerI8:-2145896039) ([raw file](https://github.com/servo/servo/blob/c4ccde386bf555757194b209b1de08a7b787765f/components/layout/query.rs#L441)):*
No more Hack, much nicer!

---

*[components/layout/query.rs, line 536 \[r8\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEpYM_bgxnhQefWDgjd:-KEqKjauVbs36K_Wn6rI:1731385556) ([raw file](https://github.com/servo/servo/blob/c4ccde386bf555757194b209b1de08a7b787765f/components/layout/query.rs#L536)):*
Make this function return a `LayerId` then move the `Some` to whre `layer_id_response` is set?

---

*[components/script/dom/element.rs, line 194 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnHLHSUA-4r2Ch5p9r:-KEqLZa7y86dA4ZtCGeB:1403127255) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/script/dom/element.rs#L194)):*
I think in this case code readability wins.

---

*[components/script/dom/element.rs, line 181 \[r8\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEqL5BLzBoqY-4I4uCN:-KEqL5BLzBoqY-4I4uCO:-239519135) ([raw file](https://github.com/servo/servo/blob/c4ccde386bf555757194b209b1de08a7b787765f/components/script/dom/element.rs#L181)):*
Much cleaner! Rewrite as:
```rust
self.has_css_layout_box() &&
!self.overflow_x_is_visible() &&
!self.overflow_y_is_visible()
```
?

---

*[components/script/dom/window.rs, line 1144 \[r7\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEnHziO7SZMEwRtKvAv:-KEqM6t9vRwWYq6Ye75G:-931466486) ([raw file](https://github.com/servo/servo/blob/ef8f08eeb6e089b2941084d528fc1a124b7d80ee/components/script/dom/window.rs#L1144)):*
@danlroberson: tahnks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968#-:-KEqMDhRK99YfskAk7P1:-787549131)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: 10 of 12 files reviewed at latest revision, 6 unresolved discussions.

---

*[components/script/dom/element.rs, line 181 \[r8\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEqL5BLzBoqY-4I4uCN:-KEqPDCEwjknPSkdl4Vt:802808075) ([raw file](https://github.com/servo/servo/blob/c4ccde386bf555757194b209b1de08a7b787765f/components/script/dom/element.rs#L181)):*
Hmm, your version looks nicer

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

danlrobertson commented:
Well done good sir :tophat: 

---

Review status: 8 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[components/layout/query.rs, line 119 \[r9\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEqUXiCg6Y_ouOorRVn:-KEqUXiEx6aQJdAI8DDj:-839622778) ([raw file](https://github.com/servo/servo/blob/6e782552e9d20fa0fc729596bd56de5614230692/components/layout/query.rs#L119)):*
Why the `match`? Wouldn't the following work

```rust
fn node_overflow(&self) ->NodeOverflowResponse {
    self.0.lock().unwrap().overflow_response
    // or even the following might work
    NodeOverflowResponse(self.0.lock().unwrap().overflow_response.0)
}
```

I might be missing something. I'm learning too :smile:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968#-:-KEqWzjafG_IaEcKLxCB:1677236304)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Reviewed 1 of 6 files at r8, 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/layout/query.rs, line 119 \[r9\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEqUXiCg6Y_ouOorRVn:-KErLk2K6A_U1fhRJYpc:1461765252) ([raw file](https://github.com/servo/servo/blob/6e782552e9d20fa0fc729596bd56de5614230692/components/layout/query.rs#L119)):*
Indeed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[components/layout/query.rs, line 119 \[r9\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEqUXiCg6Y_ouOorRVn:-KEsMW2tfYEMG42irG4W:864195940) ([raw file](https://github.com/servo/servo/blob/6e782552e9d20fa0fc729596bd56de5614230692/components/layout/query.rs#L119)):*
The first one won't work because "cannot move out of borrowed content", but the second one magically works. Actually I am not very clear about why compiler won't think that I am *moving* the field rather than *copying*?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

danlrobertson commented:
Review status: 11 of 12 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[components/layout/query.rs, line 119 \[r9\]](https://reviewable.io:443/reviews/servo/servo/9968#-KEqUXiCg6Y_ouOorRVn:-KEsUutbZ8gVJhuaakNf:1576833446) ([raw file](https://github.com/servo/servo/blob/6e782552e9d20fa0fc729596bd56de5614230692/components/layout/query.rs#L119)):*
Correct me if I'm wrong, but `NodeOverflowResponse` doesn't implement `Copy`, while [`Point2D` does](https://github.com/servo/euclid/blob/master/src/point.rs#L168). By default Rust will *move* values, and we don't own `overflow_response` in this function so we can't *move* it. The second one works because we're creating a new `NodeOverflowResponse` which we do own and we *copy* the `Point2D` because `Copy` is implemented for `Point2D` so we can *move* the new `NodeOverflowResponse`.

If you want more info the [Rust Book](https://doc.rust-lang.org/book/README.html) has three great chapters on [Ownership](https://doc.rust-lang.org/book/ownership.html), [Borrowing](https://doc.rust-lang.org/book/references-and-borrowing.html), and [Lifetimes](https://doc.rust-lang.org/book/lifetimes.html).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
There are still a couple of `unimplemented!()` functions. What is their state?

Once we're okay with this, we should squash the commits before merging.

frewsxcv commented:
*[components/script/dom/element.rs, line 1503 \[r10\]](https://reviewable.io:443/reviews/servo/servo/9968#-KF5lb407Fj_ruV82n4U:-KF5lb413g0dbUd7Dr7X:-87647570) ([raw file](https://github.com/servo/servo/blob/f3cc23b8920ad249eff2d99304ff84b87ac6d273/components/script/dom/element.rs#L1503)):*
`get_root_element()` was recently renamed to `root_element()`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Because WebIDL specified `scrollTop` and `scrollLeft` as both gettable and settable, I wrote stubs for setters here. Better idea on how to handle this?

---

Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968#-:-KF6uRkl1R7POXEyU5Fe:1904041017)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
A discussion on IRC: http://logs.glob.uno/?c=mozilla%23servo#c405321

Summary: it may be worth submitting a separate PR implementing scrolling, which would make implementing the setters much easier.

izgzhen committed:
element scroll setters

izgzhen commented:
According to my initial experiments, the code of setters depend on some code in this PR greatly. And I decide to keep building on this foundation to save trouble. Currently, the basic setters interfaces are implemented and open to review. You may see the latest commit for changes.

Basically speaking, I reused the scrolling mechanism previously used to implement `scroll` on `Window` (See ScriptToCompositorMsg::ScrollFragmentPoint`). I also wrote a WPT test combining the setters and getters.

Here are some issues unresolved yet:

First, In [step 10](https://drafts.csswg.org/cssom-view/#dom-element-scroll), I have to check the existence of [scrolling box](https://drafts.csswg.org/cssom-view/#scrolling-box) and to check if *an element has no overflow*. My initial implementation is based on my understanding of the spec, but it is showed to be wrong according to test. We should discuss this later.

Second, my investigation showed that we haven't implement the smooth scrolling yet. So the behaviour argument sent to compositor is actually just ignored. This might be an issue left to be solved in future.

Third, in my test, I intentionally made a FAIL case on the size problem (One border-width long difference from other browser). I assume that the `scroll_offset` implemented in compositor previously has subtle bug waiting to be solved.

danlrobertson commented:
I gotta say, I'm impressed... There were a few spots where I think you forgot to terminate the steps that I made notes of. I'll take another longer look later.

---

Review status: 2 of 15 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[components/script/dom/element.rs, line 1266 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFOvQnA6zJm6-UYWGHS:-KFOvQnBtQCtTCCSUpHv:1886843889) ([raw file](https://github.com/servo/servo/blob/dd60b4b1601f2f764297f62bbe2960028ed9d7ae/components/script/dom/element.rs#L1266)):*
This doesn't terminate the steps after scrolling the window. Either add a return or bundle 7, 9, and 11 into a `match`or `if-else`

---

*[components/script/dom/element.rs, line 1646 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFOwIk3X3hztma3QKBe:-KFOwIk5MsXcOFwL_-MH:1662569705) ([raw file](https://github.com/servo/servo/blob/dd60b4b1601f2f764297f62bbe2960028ed9d7ae/components/script/dom/element.rs#L1646)):*
Same here.

---

*[components/script/dom/element.rs, line 1730 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFOwRZHoJ86cHA_zOLp:-KFOwRZIhk36BxdRXx1C:1662569705) ([raw file](https://github.com/servo/servo/blob/dd60b4b1601f2f764297f62bbe2960028ed9d7ae/components/script/dom/element.rs#L1730)):*
Same here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968#-:-KFOwn4fhgEsPp5JlHCA:-1073823193)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
I'm impressed too! That's the whole scrolling API, isn't it? The PR probably needs renamed!

---

Reviewed 5 of 8 files at r11, 4 of 7 files at r12, 3 of 4 files at r13, 1 of 1 files at r14.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/compositing/compositor.rs, line 723 \[r14\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFfGoetMxTRIRZAXBYz:-KFfGoetMxTRIRZAXBZ-:-2032235281) ([raw file](https://github.com/servo/servo/blob/67276ce1b4e44e748c2dc7cf6154023e530c1ebd/components/compositing/compositor.rs#L723)):*
In the None case, should you send a message back?

---

*[components/script/dom/element.rs, line 1266 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFOvQnA6zJm6-UYWGHS:-KFfIFvKUAfuMKGt4fLi:-2091459301) ([raw file](https://github.com/servo/servo/blob/dd60b4b1601f2f764297f62bbe2960028ed9d7ae/components/script/dom/element.rs#L1266)):*
Step 7 might be better as:
```rust
        if *self.root_element() == *self {
            if doc.quirks_mode() != Quirks {
                win.scroll(x, y, behavior);
            }
            return;
        }
```
We should do something to make sure that this doesn't cause an infinite loop, perhaps adding a test `self == win`?

---

*[components/script/dom/element.rs, line 1646 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFOwIk3X3hztma3QKBe:-KFfJAvTw_sz9_RSa7gI:1700997001) ([raw file](https://github.com/servo/servo/blob/dd60b4b1601f2f764297f62bbe2960028ed9d7ae/components/script/dom/element.rs#L1646)):*
Ditto.

---

*[components/script/dom/element.rs, line 1730 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFOwRZHoJ86cHA_zOLp:-KFfJEeXBpW8pwF2sFkD:1700997001) ([raw file](https://github.com/servo/servo/blob/dd60b4b1601f2f764297f62bbe2960028ed9d7ae/components/script/dom/element.rs#L1730)):*
Ditto.

---

*[tests/wpt/web-platform-tests/cssom-view/elementScroll.html, line 1 \[r14\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFfJYOAFpI3xKom3x2R:-KFfJYOAFpI3xKom3x2S:1828065195) ([raw file](https://github.com/servo/servo/blob/67276ce1b4e44e748c2dc7cf6154023e530c1ebd/tests/wpt/web-platform-tests/cssom-view/elementScroll.html#L1)):*
Yay, more tests!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968#-:-KFfJdz857-JmnPK8X39:-1435652000)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Renamed, hope it looks better

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968#-:-KFgsu371Oxg8-4kuCdu:409430178)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/compositing/compositor.rs, line 723 \[r14\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFfGoetMxTRIRZAXBYz:-KFgt7jGLFRTRhzOd7mo:110943179) ([raw file](https://github.com/servo/servo/blob/67276ce1b4e44e748c2dc7cf6154023e530c1ebd/components/compositing/compositor.rs#L723)):*
I think that might complicate the interface. Maybe I should use a `warn` or panic like other branches in message handler?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/script/dom/element.rs, line 1266 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFOvQnA6zJm6-UYWGHS:-KFgu6wxwtust1cSkQ9L:1544159874) ([raw file](https://github.com/servo/servo/blob/dd60b4b1601f2f764297f62bbe2960028ed9d7ae/components/script/dom/element.rs#L1266)):*
An infinite loop? Do you mean that `root_element` might not terminate?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Getting there!

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[components/compositing/compositor.rs, line 723 \[r14\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFfGoetMxTRIRZAXBYz:-KFknvwz15Ghn0JP88PE:-1392737791) ([raw file](https://github.com/servo/servo/blob/67276ce1b4e44e748c2dc7cf6154023e530c1ebd/components/compositing/compositor.rs#L723)):*
You should at least emit a warning.

---

*[components/script/dom/element.rs, line 1266 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFOvQnA6zJm6-UYWGHS:-KFkoCpW39qbMY33UpCM:291648685) ([raw file](https://github.com/servo/servo/blob/dd60b4b1601f2f764297f62bbe2960028ed9d7ae/components/script/dom/element.rs#L1266)):*
We're defining `scroll`, and calling it recursively on the window, we need to make sure we're not calling it on the same object!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968#-:-KFkoUIARRFrbsv6GwJ2:-1894271234)*
<!-- Sent from Reviewable.io -->

izgzhen commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[components/script/dom/element.rs, line 1266 \[r13\]](https://reviewable.io:443/reviews/servo/servo/9968#-KFOvQnA6zJm6-UYWGHS:-KFkx6zjzF-Lrb9Rt_tV:-192028785) ([raw file](https://github.com/servo/servo/blob/dd60b4b1601f2f764297f62bbe2960028ed9d7ae/components/script/dom/element.rs#L1266)):*
Thanks for the reminder. The `scroll` is specific to window as an internal interface, logically it shouldn't call DOM API again (as I suppose), and practically the code didn't call DOM API either. This invocation is converging so I think there is no need to check.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
Reviewed 4 of 4 files at r15.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968)*
<!-- Sent from Reviewable.io -->

asajeffrey commented:
OK, I'm happy, ready for an r+? @danlrobertson look good to you?

izgzhen commented:
Great, thanks for helping the review these days

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/9968#-:-KFlp7id6xsWcAdHTyXY:324462876)*
<!-- Sent from Reviewable.io -->

danlrobertson commented:
Looks great! Excited to see what homu has to say!

asajeffrey commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 11b12f6 has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey 11b12f677b14516e92b02c0ded30310b6d6a7f96 -->

bors-servo commented:
:hourglass: Testing commit 11b12f6 with merge 8d988f2...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/394), [arm32](http://build.servo.org/builders/arm32/builds/420), [arm64](http://build.servo.org/builders/arm64/builds/410), [linux-dev](http://build.servo.org/builders/linux-dev/builds/416), [linux-rel](http://build.servo.org/builders/linux-rel/builds/457), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/430), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/420), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/432)

peterjoel commented:
@izgzhen This is some really impressive work!
